### DISCUSSION TITLE
PINN Model for an Adsorption Column for Carbon Capture

### DISCUSSION THREAD

**Discussion by Rahulx084** [lululxvi/deepxde#1325] [General]
Hello all, I'm trying to make a PINN model for the adsorption column. The feed consists of Carbon dioxide and Nitrogen. I want to generate the breakthrough curve through the PINN model. Here are the details about the Physics of the problem. It has mass and momentum balance. I've non-dimensionalized the equations. Its a 1-D equation , Carbon dioxide is the only adsorbing component. All the units are in SI unit.
This PDF have the equations and constraints written:
[PINN( Carbon Capture).pdf](https://github.com/lululxvi/deepxde/files/11530562/PINN.Carbon.Capture.pdf)

I've written the code, but unfortunately, I'm not able to get the correct breakthrough curve. The model doesn't seem to converge properly. I've tried loss weights, but still doesn't seem to give good results. Please if someone can look into it.

Here's the code

```
import deepxde as dde
import numpy as np
import tensorflow as tf

geom=dde.geometry.Interval(0,1)
timedomain=dde.geometry.TimeDomain(0,3)
geomtime=dde.geometry.GeometryXTime(geom,timedomain)

e=0.56 #bed porosity
rp=0.325/10**3 #Particle radius
rho_p=1818  #Bed density
yo=0.15 #Feed mole fraction of Carbon dioxide
L=0.2 #Bed Length
uo=0.019  #Feed velocity
k=0.0137  #Mass Transfer Coefficient (1/s)
Po=100000 #Feed Pressure
vis=0.0000172 #Gas Viscosity
Co=6.03 #Carbon dioxide concentration in feed (mol/m3)
b=(26.935/10**3) # Isotherm Constant
qsb=(5/26.935) #Bed Saturation Concentration (mol/kg)
Pe=147 #Peclet number



b_new=b*Co
m=((4*rp*rp*e*Po)/(150*(1-e)*vis*uo*L))
n=(((1-e)*rho_p*qsb)/(e*Co))
k_new=k*(L/uo)

def pde(inp,out):
    
    out_c, out_q, out_p , out_u = out[:,0:1], out[:,1:2], out[:,2:3] ,out[:,3:]  #c,q,p,u
    
  
    out_c_inp=tf.gradients(out_c,inp)[0]
    out_q_inp=tf.gradients(out_q,inp)[0]
    out_p_inp=tf.gradients(out_p,inp)[0]
    out_u_inp=tf.gradients(out_u,inp)[0]
    
    
    
    out_c_z, out_c_t= out_c_inp[:,0:1] , out_c_inp[:,1:2]   #dc/dz , dc/dt
    
    out_q_z, out_q_t= out_q_inp[:,0:1] , out_q_inp[:,1:2]
    
    out_p_z, out_p_t= out_p_inp[:,0:1] , out_p_inp[:,1:2]
    
    out_u_z= out_u_inp[:,0:1]

    out_c_zz= tf.gradients(out_c_z,inp)[0][:,0:1]   #d2c/dz2
    
    out_p_zz= tf.gradients(out_p_z,inp)[0][:,0:1]
    

    
   
    
    
    
  
    
    
    
    qe= ((b_new*out_c)/(1+b_new*out_c)) #Isotherm
    
    
    
    
    
   
    
    
    r1= out_c_t + out_u*out_c_z+ out_c*out_u_z + n*out_q_t - (1/Pe)*out_c_zz
    
    r2= out_q_t - k_new*qe + k_new*out_q
    
    r3= out_p_t +out_u*out_p_z+ out_p*out_u_z + yo*n*out_q_t - (1/Pe)*out_p_zz
    
    r4= out_u + m*out_p_z #velocity Darcy's law
     
    
    return [r1,r2,r3,r4]



def boundary_l(inp, on_boundary):
    return on_boundary and np.isclose(inp[0], 0)

def boundary_r(inp, on_boundary):
    return on_boundary and np.isclose(inp[0],1)

bc1 = dde.DirichletBC(geomtime, lambda inp:1, boundary_l,component=0)

#bc1=dde.icbc.OperatorBC(geomtime,lambda inp, out, _ :(1/Pe)*dde.grad.jacobian(out, inp, i=0, j=0)- out[:,0:1] + 1, boundary_l)

bc2 = dde.DirichletBC(geomtime, lambda inp: 1.01, boundary_r,component=2)

bc3= dde.DirichletBC(geomtime, lambda inp: 1, boundary_l,component=3)

bc4 = dde.icbc.NeumannBC(geomtime, lambda inp:0, boundary_r,component=0)

bc5 = dde.icbc.NeumannBC(geomtime, lambda inp:0, boundary_r,component=2)

ic1 = dde.icbc.IC(geomtime, lambda inp: 0, lambda _, on_initial: on_initial,component=0)

ic2 = dde.icbc.IC(geomtime, lambda inp: 0, lambda _, on_initial: on_initial,component=1)


data = dde.data.TimePDE(geomtime, pde, [bc1,bc2,bc3,bc4,bc5,ic1,ic2],
                        num_domain=30000, num_boundary=20000, num_initial=20000,num_test=3000)

net = dde.nn.FNN([2] + 6*[50] + [4], "tanh", "Glorot normal")

model =dde.Model(data, net)
model.compile("adam", lr=1e-3)
losshistory, train_state = model.train(iterations=15000,batch_size=300,model_save_path="CC_Carboncappc")




model.compile("L-BFGS")
losshistory, train_state = model.train()





dde.saveplot(losshistory, train_state, issave=True, isplot=True)







import matplotlib.pyplot as plt

for x_new in [0,0.1,0.2,0.5,1]:
    t_new=np.linspace(0,2.5,50)
    X_new,T_new=np.meshgrid(x_new,t_new)
    X_star_new = np.hstack((X_new.flatten()[:, None],T_new.flatten()[:, None]))
    C_new = model.predict(X_star_new,operator=None)
    plt.plot(t_new*10.52,C_new[:,0:1]*6.03)
plt.plot(T*10.52,C*6.03)
plt.xlabel("Time(s)")
plt.ylabel("Concentraion(mol/m3)")
plt.legend(["0.0","0.1","0.2","0.5","1"])   
plt.show(block=True)




plt.savefig ('Test2.pdf',dpi =300 , format ='pdf')



```
Here's how the loss looks like: 
```
Step      Train loss                                                                                                        Test loss                                                                                                         Test metric
0         [2.49e+01, 1.08e-02, 5.50e-01, 2.14e+06, 8.46e-01, 8.05e-01, 2.02e+00, 3.71e-03, 6.12e-02, 5.37e-03, 2.35e-02]    [1.68e+01, 7.86e-03, 3.22e-01, 1.96e+06, 8.46e-01, 8.05e-01, 2.02e+00, 3.71e-03, 6.12e-02, 5.37e-03, 2.35e-02]    []  
1000      [5.37e-02, 1.03e-03, 4.46e-02, 1.19e+01, 1.01e-01, 7.60e-02, 1.18e+00, 7.73e-03, 3.53e-07, 1.83e-02, 3.36e-04]    [7.19e-02, 1.38e-03, 4.16e-02, 1.08e+01, 1.01e-01, 7.60e-02, 1.18e+00, 7.73e-03, 3.53e-07, 1.83e-02, 3.36e-04]    []  
2000      [4.34e-03, 9.34e-04, 4.74e-02, 4.74e+00, 8.62e-02, 8.99e-03, 4.51e-01, 5.49e-03, 1.58e-07, 1.61e-02, 9.29e-05]    [6.84e-03, 1.18e-03, 4.77e-02, 5.32e+00, 8.62e-02, 8.99e-03, 4.51e-01, 5.49e-03, 1.58e-07, 1.61e-02, 9.29e-05]    []  
3000      [2.50e-03, 8.19e-04, 2.64e-02, 2.02e+00, 7.42e-02, 4.28e-03, 1.50e-01, 6.72e-03, 7.75e-08, 1.38e-02, 3.92e-05]    [4.18e-03, 9.79e-04, 2.92e-02, 2.58e+00, 7.42e-02, 4.28e-03, 1.50e-01, 6.72e-03, 7.75e-08, 1.38e-02, 3.92e-05]    []  
4000      [2.30e-03, 7.35e-04, 1.42e-02, 1.24e+00, 6.46e-02, 3.46e-03, 9.89e-02, 7.86e-03, 6.77e-08, 1.25e-02, 2.14e-05]    [3.84e-03, 8.27e-04, 1.71e-02, 1.60e+00, 6.46e-02, 3.46e-03, 9.89e-02, 7.86e-03, 6.77e-08, 1.25e-02, 2.14e-05]    []  
5000      [2.35e-03, 6.86e-04, 6.49e-03, 1.08e+00, 5.87e-02, 1.92e-03, 8.11e-02, 5.79e-03, 6.73e-08, 1.15e-02, 1.59e-05]    [3.52e-03, 7.12e-04, 8.37e-03, 1.25e+00, 5.87e-02, 1.92e-03, 8.11e-02, 5.79e-03, 6.73e-08, 1.15e-02, 1.59e-05]    []  
6000      [1.70e-03, 6.60e-04, 4.28e-03, 6.35e-01, 5.53e-02, 1.08e-03, 5.80e-02, 2.96e-03, 4.96e-08, 1.09e-02, 1.40e-05]    [2.54e-03, 6.44e-04, 5.46e-03, 8.29e-01, 5.53e-02, 1.08e-03, 5.80e-02, 2.96e-03, 4.96e-08, 1.09e-02, 1.40e-05]    []  
7000      [1.01e-03, 6.43e-04, 2.72e-03, 4.70e-01, 5.08e-02, 3.20e-04, 3.84e-02, 1.75e-03, 4.30e-08, 1.10e-02, 1.65e-05]    [1.49e-03, 5.95e-04, 2.95e-03, 5.95e-01, 5.08e-02, 3.20e-04, 3.84e-02, 1.75e-03, 4.30e-08, 1.10e-02, 1.65e-05]    []  
8000      [5.66e-04, 6.44e-04, 2.74e-03, 6.15e-01, 4.78e-02, 1.19e-04, 2.65e-02, 1.04e-03, 4.35e-08, 1.05e-02, 3.55e-05]    [6.14e-04, 5.54e-04, 1.21e-03, 5.19e-01, 4.78e-02, 1.19e-04, 2.65e-02, 1.04e-03, 4.35e-08, 1.05e-02, 3.55e-05]    []  
9000      [2.41e-04, 6.56e-04, 2.49e-03, 2.49e-01, 4.50e-02, 3.03e-04, 1.12e-02, 1.06e-03, 3.50e-08, 1.08e-02, 3.53e-05]    [3.14e-04, 5.56e-04, 9.38e-04, 2.89e-01, 4.50e-02, 3.03e-04, 1.12e-02, 1.06e-03, 3.50e-08, 1.08e-02, 3.53e-05]    []  
10000     [2.29e-04, 6.80e-04, 2.15e-03, 1.24e-01, 4.32e-02, 5.41e-04, 4.96e-03, 1.06e-03, 3.29e-08, 1.09e-02, 3.33e-05]    [3.53e-04, 5.73e-04, 8.52e-04, 1.61e-01, 4.32e-02, 5.41e-04, 4.96e-03, 1.06e-03, 3.29e-08, 1.09e-02, 3.33e-05]    []  
11000     [3.88e-04, 7.09e-04, 3.27e-03, 6.88e-02, 4.34e-02, 1.17e-03, 1.89e-03, 1.21e-03, 3.12e-08, 1.00e-02, 2.86e-05]    [5.57e-04, 5.72e-04, 9.42e-04, 8.51e-02, 4.34e-02, 1.17e-03, 1.89e-03, 1.21e-03, 3.12e-08, 1.00e-02, 2.86e-05]    []  
12000     [4.87e-04, 7.32e-04, 4.87e-03, 3.78e-02, 4.40e-02, 1.88e-03, 1.21e-03, 1.56e-03, 3.22e-08, 9.27e-03, 2.79e-05]    [6.74e-04, 5.65e-04, 1.38e-03, 4.28e-02, 4.40e-02, 1.88e-03, 1.21e-03, 1.56e-03, 3.22e-08, 9.27e-03, 2.79e-05]    []  
13000     [2.92e-04, 7.58e-04, 4.55e-03, 2.12e-02, 4.34e-02, 2.22e-03, 1.52e-03, 1.46e-03, 3.49e-08, 9.02e-03, 3.03e-05]    [3.58e-04, 5.83e-04, 1.48e-03, 2.67e-02, 4.34e-02, 2.22e-03, 1.52e-03, 1.46e-03, 3.49e-08, 9.02e-03, 3.03e-05]    []  
14000     [2.41e-04, 8.04e-04, 5.21e-03, 1.45e-02, 4.29e-02, 2.31e-03, 2.10e-03, 1.58e-03, 3.64e-08, 8.45e-03, 3.27e-05]    [2.63e-04, 5.98e-04, 1.93e-03, 1.61e-02, 4.29e-02, 2.31e-03, 2.10e-03, 1.58e-03, 3.64e-08, 8.45e-03, 3.27e-05]    []  
15000     [1.84e-03, 8.20e-04, 5.52e-03, 1.69e+00, 4.20e-02, 2.07e-03, 2.99e-03, 1.12e-03, 5.80e-09, 9.19e-03, 4.40e-05]    [8.22e-04, 6.08e-04, 2.38e-03, 2.54e+00, 4.20e-02, 2.07e-03, 2.99e-03, 1.12e-03, 5.80e-09, 9.19e-03, 4.40e-05]    []  

```
